---
title: "Data preprocessing for TRIPOD - 10x Genomics PBMC 10k"
author: "Yuchao Jiang, Yuriko Harigaya, Nancy R. Zhang"
date: "`r format(Sys.Date())`"
output:
    html_document:
        highlight: pygments
        toc: true
bibliography: TRIPOD.bib
vignette: >
  %\VignetteIndexEntry{preprocessing}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE, warning = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# 1 Introduction

This is a supplement to [the TRIPOD vignette for the 10x Genomics PBMC data](http://htmlpreview.github.io/?https://github.com/yuchaojiang/TRIPOD/blob/main/vignettes/TRIPOD_pbmc.html).
Here, we demonstrate data preprocessing required for conducting
analysis using the TRIPOD package.
The input is sigle-cell ATAC/RNA multiomic data of human peripheral blood 
mononuclear cells (PBMCs), which is publicly available from 
[10x Genomics](https://support.10xgenomics.com/single-cell-multiome-atac-gex/datasets). 

# 2 Required packages

The TRIPOD package is available from [GitHub](https://github.com/yuchaojiang/TRIPOD).

```{r install, eval = FALSE, message = FALSE, warning = FALSE}
if (!requireNamespace("devtools", quietly = TRUE)) install.packages("devtools")
devtools::install_github("yuchaojiang/TRIPOD/package")
```

```{r setup, eval = TRUE, message = FALSE, warning=FALSE}
library(TRIPOD)
```

Additionally, the following packages are required to run this vignette.

```{r packages, cache = FALSE, message = FALSE}
library(BiocParallel)
library(dplyr)
library(ggplot2)
library(patchwork)
library(Seurat)
library(SeuratDisk)
library(Signac)
library(GenomicRanges)
library(GenomeInfoDb)
library(EnsDb.Hsapiens.v86)
library(BSgenome.Hsapiens.UCSC.hg38)
library(chromVAR)
library(JASPAR2020)
library(TFBSTools)
library(motifmatchr)
```

# 3 Preprocessing
## 3.1 Creating and processing a Seurat object

The input data can be downlowded from below.

```{sh, eval=FALSE}
wget https://cf.10xgenomics.com/samples/cell-arc/1.0.0/pbmc_granulocyte_sorted_10k/pbmc_granulocyte_sorted_10k_filtered_feature_bc_matrix.h5
wget https://cf.10xgenomics.com/samples/cell-arc/1.0.0/pbmc_granulocyte_sorted_10k/pbmc_granulocyte_sorted_10k_atac_fragments.tsv.gz
wget https://cf.10xgenomics.com/samples/cell-arc/1.0.0/pbmc_granulocyte_sorted_10k/pbmc_granulocyte_sorted_10k_atac_fragments.tsv.gz.tbi
```

We first create and process a Seurat object essentially according to the tutorial [here](https://satijalab.org/seurat/articles/weighted_nearest_neighbor_analysis.html). 

```{r, eval=TRUE, warning=FALSE,  cache = FALSE, message = FALSE, fig.width = 9, fig.height = 4, fig.align = "center"}
inputdata.10x <- Read10X_h5("pbmc_granulocyte_sorted_10k_filtered_feature_bc_matrix.h5")
rna_counts <- inputdata.10x$`Gene Expression`
atac_counts <- inputdata.10x$Peaks

pbmc <- CreateSeuratObject(counts = rna_counts)
pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^MT-")

grange.counts <- StringToGRanges(rownames(atac_counts), sep = c(":", "-"))
grange.use <- seqnames(grange.counts) %in% standardChromosomes(grange.counts)
atac_counts <- atac_counts[as.vector(grange.use), ]
annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v86)
seqlevelsStyle(annotations) <- 'UCSC'
genome(annotations) <- "hg38"

frag.file <- "pbmc_granulocyte_sorted_10k_atac_fragments.tsv.gz"
chrom_assay <- CreateChromatinAssay(
  counts = atac_counts,
  sep = c(":", "-"),
  genome = 'hg38',
  fragments = frag.file,
  min.cells = 10,
  annotation = annotations
)
pbmc[["ATAC"]] <- chrom_assay

pbmc <- subset(x = pbmc,
    subset = nCount_ATAC < 7e4 &
    nCount_ATAC > 5e3 &
    nCount_RNA < 25000 &
    nCount_RNA > 1000 &
    percent.mt < 20)

DefaultAssay(pbmc) <- "RNA"
pbmc <- SCTransform(pbmc, verbose = FALSE) %>% RunPCA(verbose = FALSE) %>% 
  RunUMAP(dims = 1:50, reduction.name = 'umap.rna', reduction.key = 'rnaUMAP_',
          verbose = FALSE)

DefaultAssay(pbmc) <- "ATAC"
pbmc <- RunTFIDF(pbmc)
pbmc <- FindTopFeatures(pbmc, min.cutoff = 'q0')
pbmc <- RunSVD(pbmc)
pbmc <- RunUMAP(pbmc, reduction = 'lsi', dims = 2:50, reduction.name = "umap.atac", 
                reduction.key = "atacUMAP_")

pbmc <- FindMultiModalNeighbors(pbmc, reduction.list = list("pca", "lsi"), dims.list = list(1:50, 2:50))
pbmc <- RunUMAP(pbmc, nn.name = "weighted.nn", reduction.name = "wnn.umap", reduction.key = "wnnUMAP_")
pbmc <- FindClusters(pbmc, graph.name = "wsnn", algorithm = 3, verbose = FALSE)

# Run chromVAR to generate the motif deviation scores
DefaultAssay(pbmc) <- "ATAC"
pwm_set <- getMatrixSet(x = JASPAR2020, opts = list(species = 9606, all_versions = FALSE))
motif.matrix <- CreateMotifMatrix(features = granges(pbmc), pwm = pwm_set, genome = 'hg38', use.counts = FALSE)
motif.object <- CreateMotifObject(data = motif.matrix, pwm = pwm_set)

pbmc <- SetAssayData(pbmc, assay = 'ATAC', slot = 'motifs', new.data = motif.object)

pbmc <- RunChromVAR(
  object = pbmc,
  genome = BSgenome.Hsapiens.UCSC.hg38
)
```

We obtain cell type labels according to the tutorial
[here](https://satijalab.org/seurat/articles/multimodal_reference_mapping.html).
The input CITE-seq data can be downloaded from [here](https://atlas.fredhutch.org/data/nygc/multimodal/pbmc_multimodal.h5seurat).

```{r, eval=TRUE, warning=FALSE,  cache = FALSE, message = FALSE, fig.width = 9, fig.height = 4, fig.align = "center"}
reference <- LoadH5Seurat("pbmc_multimodal.h5seurat")

p1=DimPlot(object = reference, 
           reduction = "wnn.umap",
           group.by = "celltype.l1", 
           label = TRUE, 
           label.size = 3, 
           repel = TRUE) + NoLegend() + ggtitle("CITE-seq celltype l1")
p2=DimPlot(object = reference, 
           reduction = "wnn.umap",
           group.by = "celltype.l2", 
           label = TRUE, 
           label.size = 3, 
           repel = TRUE) + NoLegend() + ggtitle("CITE-seq celltype l2")
p1+p2

# Find anchors between reference and query
# using a precomputed supervised PCA (spca) transformation for this example. 
DefaultAssay(pbmc) <- "SCT"
anchors <- FindTransferAnchors(
  reference = reference,
  query = pbmc,
  normalization.method = "SCT",
  reference.reduction = "spca",
  dims = 1:50,
  recompute.residuals = FALSE
)

# Transfer cell type labels and protein data from the reference to the query
# project the query data onto the umap structure of the reference
pbmc <- MapQuery(
  anchorset = anchors,
  query = pbmc,
  reference = reference,
  refdata = list(
    celltype.l1 = "celltype.l1",
    celltype.l2 = "celltype.l2",
    predicted_ADT = "ADT"
  ),
  reference.reduction = "spca", 
  reduction.model = "wnn.umap"
)

pbmc$celltype=as.factor(pbmc$predicted.celltype.l2)

# Remove some minor cell types (with less than 100 cells)
pbmc$celltype[pbmc$celltype=='cDC2']='cDC'
pbmc=pbmc[,!pbmc$celltype%in%names(which(table(pbmc$celltype)<100))]
pbmc$celltype=droplevels(pbmc$celltype)

# Renormalize after QC
DefaultAssay(pbmc) <- "RNA"
pbmc <- SCTransform(pbmc, verbose = FALSE) %>% RunPCA(verbose = FALSE) %>% 
  RunUMAP(dims = 1:50, reduction.name = 'umap.rna', reduction.key = 'rnaUMAP_', verbose = FALSE)

DefaultAssay(pbmc) <- "ATAC"
pbmc <- RunTFIDF(pbmc)
pbmc <- FindTopFeatures(pbmc, min.cutoff = 'q0')
pbmc <- RunSVD(pbmc)
pbmc <- RunUMAP(pbmc, reduction = 'lsi', dims = 2:50, reduction.name = "umap.atac", reduction.key = "atacUMAP_")

pbmc <- FindMultiModalNeighbors(pbmc, reduction.list = list("pca", "lsi"), dims.list = list(1:50, 2:50))
pbmc <- RunUMAP(pbmc, nn.name = "weighted.nn", reduction.name = "wnn.umap", reduction.key = "wnnUMAP_")
pbmc <- FindClusters(pbmc, graph.name = "wsnn", algorithm = 3, verbose = FALSE)

# For each cluster, we only keep the main cell type
cell.keep=rep(FALSE, ncol(pbmc))
temp=table(pbmc$seurat_clusters, pbmc$celltype)
for(i in 1:nrow(temp)){
  clusteri=as.numeric(rownames(temp)[i])
  celltypei=colnames(temp)[which.max(temp[i,])]
  cell.keep[which(pbmc$seurat_clusters==clusteri & pbmc$celltype==celltypei)]=TRUE
}
pbmc=pbmc[,cell.keep]

# Renormalize after QC
DefaultAssay(pbmc) <- "RNA"
pbmc <- SCTransform(pbmc, verbose = FALSE) %>% RunPCA(verbose = FALSE) %>% 
  RunUMAP(dims = 1:50, reduction.name = 'umap.rna', reduction.key = 'rnaUMAP_', 
          verbose = FALSE)

DefaultAssay(pbmc) <- "ATAC"
pbmc <- RunTFIDF(pbmc)
pbmc <- FindTopFeatures(pbmc, min.cutoff = 'q0')
pbmc <- RunSVD(pbmc)
pbmc <- RunUMAP(pbmc, reduction = 'lsi', dims = 2:50, reduction.name = "umap.atac", reduction.key = "atacUMAP_")

pbmc <- FindMultiModalNeighbors(pbmc, reduction.list = list("pca", "lsi"), dims.list = list(1:50, 2:50))
pbmc <- RunUMAP(pbmc, nn.name = "weighted.nn", reduction.name = "wnn.umap", reduction.key = "wnnUMAP_")
pbmc <- FindClusters(pbmc, graph.name = "wsnn", algorithm = 3, verbose = FALSE)

p1 <- DimPlot(pbmc, reduction = "umap.rna", group.by = "predicted.celltype.l1", 
              label = TRUE, label.size = 3, repel = TRUE) + ggtitle("RNA")
p2 <- DimPlot(pbmc, reduction = "umap.atac", group.by = "predicted.celltype.l1", 
              label = TRUE, label.size = 3, repel = TRUE) + ggtitle("ATAC")
p3 <- DimPlot(pbmc, reduction = "wnn.umap", group.by = "predicted.celltype.l1", 
              label = TRUE, label.size = 3, repel = TRUE) + ggtitle("WNN")
p1 + p2 + p3 & NoLegend() & theme(plot.title = element_text(hjust = 0.5))

p1 <- DimPlot(pbmc, reduction = "umap.rna", group.by = "predicted.celltype.l2", 
              label = TRUE, label.size = 3, repel = TRUE) + ggtitle("RNA")
p2 <- DimPlot(pbmc, reduction = "umap.atac", group.by = "predicted.celltype.l2", 
              label = TRUE, label.size = 3, repel = TRUE) + ggtitle("ATAC")
p3 <- DimPlot(pbmc, reduction = "wnn.umap", group.by = "predicted.celltype.l2", 
              label = TRUE, label.size = 3, repel = TRUE) + ggtitle("WNN")
p1 + p2 + p3 & NoLegend() & theme(plot.title = element_text(hjust = 0.5))
```

## 3.2 Creating objects required for model fitting

From the processed Seurat object, we create a list object containing 
information that is necessary for model fitting, 
namely, `transcripts.gr`, `peaks.gr`, `motifxTF`, and `peakxmotif`,
using the `getObjectsForModelFit()` function. 
See the documentation of the function for descriptions of these objects.
Note the following points.

* We only keep motifs when genes encoding corresponding TFs are present in the
RNA-seq data.
* We ensure that the same set of genes is included in all objects. 

```{r, eval = TRUE, message = FALSE, warning = FALSE}
tripod.pbmc <- getObjectsForModelFit(object = pbmc, chr = paste0("chr", 1:22))
```

Individual objects can be extracted using the following code.

```{r, eval = TRUE, message = FALSE, warning = FALSE}
transcripts.gr <- tripod.pbmc$transcripts.gr
peaks.gr <- tripod.pbmc$peaks.gr
motifxTF <- tripod.pbmc$motifxTF
peakxmotif <- tripod.pbmc$peakxmotif
```

After this step, the Seurat object needs to be filtered.

```{r, eval = TRUE, message = FALSE, warning = FALSE}
pbmc <- filterSeuratObject(object = pbmc, tripod.object = tripod.pbmc)
```

We note that our current implementation
excludes cases where a TF corresponds to multiple motifs or vice versa. 
We plan to provide an option to include such degenerate cases in a future version
of the package.

## 3.3 Reprocessing of the Seurat object

We then perform normalization, dimension reduction, nearest neighbor graph construction,
and Uniform Manifold Approximation and Projection (UMAP) embedding for the filtered data
using the `processSeuratObject()` function.

```{r renormalization, eval = TRUE, message = FALSE, warning = FALSE}
pbmc <- processSeuratObject(object = pbmc, dim.rna = 1:50, dim.atac = 2:50,
	verbose = FALSE)
```

The following code visualizes the data on recalculated graphs.

```{r, eval = TRUE, message = FALSE, warning = FALSE, fig.width = 9, fig.height = 4, fig.align = "center"}
p1 <- DimPlot(pbmc, reduction = "umap.rna",  group.by = "celltype", 
	label = TRUE, label.size = 2.5, repel = TRUE) +
	ggtitle("RNA")
p2 <- DimPlot(pbmc, reduction = "umap.atac",  group.by = "celltype", 
	label = TRUE, label.size = 2.5, repel = TRUE) +
	ggtitle("ATAC")
p3 <- DimPlot(pbmc, reduction = "wnn.umap", group.by = "celltype", 
	label = TRUE, label.size = 2.5, repel = TRUE) +
	ggtitle("WNN")
p1 + p2 + p3 & NoLegend() & theme(plot.title = element_text(hjust = 0.5))
```

## 3.4 Partitioning single cells into metacells

To alleviate the issue of sparsity of the single cell data, we next partition the cells
into clusters, which we call metacells. The models in the TRIPOD package are
based on metacells, not single cells. Cells can be clustered based on either the ATAC
or RNA modality. It is also possible to use the weighted nearest neighbor (WNN) method [@Hao2021].
For the PBMC data, we use the WNN method.

To obtain a set of metacells suited for model fitting, it is necessary to select
an optimal resolution of clustering to ensure the following.

* a sufficient number of cells per metacell for resolving the sparsity issue
* a sufficient number of metacells to allow for model fitting

Our empirical approach is to select a resolution that gives 80 or more metacells,
the majority of which contain 20 or more single cells. For this purpose, 
based on the graph calculated in the previous section,
we then perform clustering at several resolutions.
The following code generates a data frame containing the resolutions,
the number of resultant clusters, and the number of clusters with fewer than
threshold number of single cells. The threshold is set to 20 cells.

```{r, eval = TRUE, message = FALSE, warning = FALSE}
set.seed(123)
num.clusters <- optimizeResolution(
	object = pbmc,
	graph.name = "wsnn",
	assay.name = "WNN",
	resolutions = seq(10, 35, 5),
	min.num = 20
)
num.clusters
```

Based on the above analysis, we select the resolution of 15.

```{r, eval = TRUE, message = FALSE, warning = FALSE}
res <- 15
pbmc <- getClusters(object = pbmc, graph.name = "wsnn", algorithm = 3,
    	resolution = res, verbose = FALSE)
```

We now obtain matrices containing normalized RNA expression and chromatin accessibility per metacell using the `getMetacellMatrices()` function. The function returns a list object containing metacell matrices for
the RNA and ATAC modalities.

```{r, eval = TRUE, message = FALSE, warning = FALSE}
metacell.pbmc <- getMetacellMatrices(object = pbmc,
  cluster.name = "seurat_clusters"
)
```

Individual matrices can be extracted using the following code.

```{r, eval = TRUE, message = FALSE, warning = FALSE}
metacell.rna <- metacell.pbmc$rna
metacell.peak <- metacell.pbmc$peak
```

It is necessary to save the Seurat object (`pbmc`) for running [the TRIPOD vignette](http://htmlpreview.github.io/?https://github.com/yuchaojiang/TRIPOD/blob/main/vignettes/TRIPOD_pbmc.html)
since the object is not included in the TRIPOD package.

```{r, eval = TRUE}
saveRDS(pbmc, file = 'pbmc.rds')
```

## 3.5 Assigning cell types and colors for visualization

For visualization purposes, it is convenient to assign colors to cell types
and create objects storing mapping between them.
To obtain the objects, we use the `getColors()` function,
which returns data frames containing
mappings of colors to single cells, metacells, and cell types.

```{r, eval = TRUE, message = FALSE, warning = FALSE}
pbmc$celltype=factor(pbmc$celltype)
color.pbmc <- getColors(object = pbmc)
```

The following code extracts the mapping between the metacells and cell types
as a character string. Note that the cell types are ordered by the metacell
indices.

```{r, eval = TRUE, message = FALSE, warning = FALSE}
metacell.celltype <- color.pbmc$metacell$celltype
metacell.celltype.col <- color.pbmc$metacell$color
```

## 3.6 Getting a list of target genes

We typically run model fitting for top-ranked highly variable genes (HVGs)
identified by the `VariableFeatures()` function from the Seurat packages.
The following code gives a ordered list of top 3000 HVGs.

```{r, eval = TRUE, message = FALSE, warning = FALSE}
# obtain highly variable genes based on SCT
DefaultAssay(pbmc) <- "SCT"
hvg.pbmc <- VariableFeatures(pbmc)
```

## 3.7 Save intermediate files

Here, we save the intermediate files that are needed for the next tutorial:
[TRIPOD vignette for the 10x Genomics PBMC data](http://htmlpreview.github.io/?https://github.com/yuchaojiang/TRIPOD/blob/main/vignettes/TRIPOD_pbmc.html).

```{r, eval = TRUE}
pbmc.transcripts.gr=transcripts.gr
save(pbmc.transcripts.gr, file = 'pbmc.transcripts.gr.rda')
pbmc.peaks.gr=peaks.gr
save(pbmc.peaks.gr, file='pbmc.peaks.gr.rda')
pbmc.motifxTF=motifxTF
save(pbmc.motifxTF, file='pbmc.motifxTF.rda')
pbmc.peakxmotif=peakxmotif
save(pbmc.peakxmotif, file='pbmc.peakxmotif.rda')
pbmc.metacell.rna=metacell.rna
save(pbmc.metacell.rna, file='pbmc.metacell.rna.rda')
pbmc.metacell.peak=metacell.peak
save(pbmc.metacell.peak, file='pbmc.metacell.peak.rda')
pbmc.metacell.celltype=metacell.celltype
save(pbmc.metacell.celltype, file='pbmc.metacell.celltype.rda')
pbmc.metacell.celltype.col=metacell.celltype.col
save(pbmc.metacell.celltype.col, file='pbmc.metacell.celltype.col.rda')
```

# Session information

```{r, eval = TRUE}
sessionInfo()
```

# References
