---
title: "TRIPOD - 10x Genomics mouse embryonic brain"
author: "Yuriko Harigaya, Nancy R. Zhang, Yuchao Jiang"
date: "`r format(Sys.Date())`"
output:
    html_document:
        highlight: pygments
        toc: true
vignette: >
  %\VignetteIndexEntry{TRIPOD}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# 1 Introduction

Proper regulation of gene transcription is crucial for cell identity and function,
and an accurate understanding of this process is expected to provide insights
into disease mechanisms. While the transcriptional regulation of a given gene 
involves complex interplay between multiple factors, in one view, the system
can be decomposed into modules consisting of a *cis*-regulatory DNA element and 
a transcription factor (TF) that activates or represses the target gene through
binding to the element. The recently developed single-cell multiomics technology,
in which both RNA expression and chromatin accessibility are measured
at a genome-wide level using ATAC-seq and RNA-seq in the same single cells, provides
ideal opportunities to interrogate such regulatory logic.

TRIPOD is a statistical framework for detecting three-way regulatory relationships
between a *cis*-regulatory region, a TF, and a target gene, which we call a "trio,"
using single-cell ATAC/RNA multiomic data. The main functionality of this package is:

* Infers trio regulatory relationships using robust nonparametric models

The package is augmented by the following two related functionalities:

* Builds RNA prediction models from ATAC-seq data
* Identifies cell-types in which a trio regulatory relationship is active 
by estimating the influence of a data point in linear regression

Here, we demonstrate the use of TRIPOD to study epigenetic control of 
gene expression using sigle-cell multiomic data, which is publicly available from 
[10x Genomics](https://support.10xgenomics.com/single-cell-multiome-atac-gex/datasets). 
The data was generated in 4881 cells of the mouse embryonic brain at day 18 
and contains ATAC-seq and RNA-seq read counts for 144437 ATAC peaks 
and 32285 gene regions, respectively.

# 2 Basic data formats and notations in TRIPOD

The processed ATAC-seq data from multiomic experiments can be represented as a
$c$ by $n$ matrix, $X$, where $c$ is the number of cells and $n$ is the number of peaks.
$X_{it}$ represents the $(i, t)$ element for $i = 1, \cdots, c$
and $t = 1, \cdots, n$.
The processed RNA-seq data can be represented by a $c$ by $m$ matrix, $Y$, 
where $c$ is the number of cells and $m$ is the number of genes.
$Y_{ig}$ represents the $(i, g)$ element for $i = 1, \cdots, c$
and $g = 1, \cdots, m$.
Importantly, $c$ is shared between $X$ and $Y$ because of the multiomic technology.
For a given gene $g$, we consider a set of ATAC peaks near the 
transcription start site (TSS), which are candidate *cis*-regulatory elements.
In most cases, we include peaks 100 to 200 Kb up and downstream from TSS.
We also consider RNA expression levels of a set of TFs, $Y_{ij}$, 
whose binding motifs are present in any of the peaks.
That is, we use the same RNA-sequencing data for the target gene expression
and TF activity but with different subscripts, $g$ for target genes and $j$ for TFs.
A key point is that, in our system, cell $i$ represents a group of similar cells, 
which we call a "metacell," rather than a single cell.
Signals for single cells in the multimodal data are excessively sparse, and,
to resolve the issue of the sparsity, it is necessary to aggregate information 
across cells.

# 3 Overview of the workflow

## 3.1 Data preprocessing

The data preprocessing is performed using the Seurat and Signac packages
according to the accompanying tutorials (see [the preprocessing vignette](http://htmlpreview.github.io/?https://github.com/yharigaya/TRIPOD/blob/main/vignettes/preprocessing_e18.html)).
The first step is to create a Seurat object containing both single-cell ATAC-seq and RNA-seq data. We then add TF motif information to the Seurat object.

From the Seurat object, we next extract smaller objects used for model fitting
including the following:

* Chromosomal coordinates of target genes (`transcripts.gr`)
* ATAC peaks thereof (`peaks.gr`)
* Mapping between TFs and motifs (`motifxTF`)
* Indicators as to whether a given motif exists in ATAC peaks (`peakxmotif`)

We then create metacell matrices by aggregating "similar" cells.
The groups of cells can be identified based on either ATAC or RNA modalities.
Alternatively, Weighted Nearest Neighbot (WNN) method can be used to partition
single cells into metacell clusters.

## 3.2 Model fitting

The first step in the model fitting for identifying three-way regulatory
relationship is to create a `xymats` object for a target gene of interest
(Section 5.1). `xymats` is a list containing information that is necessary to
perform model fitting across all candidate trios for the target gene, including
the $Y_g$, $X_t$, and $Y_j$ vectors (`xymats$Yg`, `xymats$Xt`, and `xymats$Yj`).
Note that the `xymats$Xt` and `xymats$Yj` elements are matrices with
rows representing metacells and
columns representing peaks and TFs, respectively.
The `xymats` object is also used as input for RNA prediction (Section 6).
For the influential analysis, we first extract the $Y_g$, $X_t$, and $Y_j$ vectors
for a given trio (Section 7).

We fit the TRIPOD model as well as other related models using the `fitModel()` 
function, which returns a `xymats` object with estimates of coefficients
and associated $p$-values.
Attention needs to be paid for the following points.

* Different models returns `xymats` objects with different elements since they
yields estimates of different sets of coefficients.
* By default, `fitModel()` caps outliers in the `xymats$Yg`, `xymats$Xt`, 
and `xymats$Yj` elements, and the elements are replaced with the capped values.
That is, the elements differ between the input and output.

The formulation of the models is described in detail in [our manuscript](https://www.biorxiv.org/content/10.1101/2021.09.22.461437v1).

# 4 Required packages

The TRIPOD package is available from [GitHub](https://github.com/yharigaya/TRIPOD).

```{r install, eval = TRUE, message = FALSE, warning = FALSE, results = FALSE}
if (!requireNamespace("devtools", quietly = TRUE)) install.packages("devtools")
devtools::install_github("yharigaya/TRIPOD/package")
```

```{r setup, eval = TRUE, message = FALSE}
library(TRIPOD)
```

Additionally, the following packages are required to run this vignette.

```{r packages, cache = FALSE, message = FALSE}
library(Seurat)
library(Signac)
library(GenomicRanges)
library(dplyr)
library(ggplot2)
library(EnsDb.Mmusculus.v79)
library(GenomeInfoDb)
library(patchwork)
library(BiocParallel)
```

# 5 Detecting candidate trio regulatory relationships

In this section, we demonstrate how to screen for trio regulatory relationships
using data objects that have been preprocessed from the multiomics data and
included in the TRIPOD package.
The preprocessing is described in [the accompanying preprocessing vignette](http://htmlpreview.github.io/?https://github.com/yharigaya/TRIPOD/blob/main/vignettes/preprocessing_e18.html).

The following code attaches all necessary objects.

```{r, eval = TRUE}
data(transcripts.gr)
data(peaks.gr)
data(motifxTF)
data(peakxmotif)
data(metacell.rna)
data(metacell.peak)
data(metacell.celltype)
data(metacell.celltype.col)
```

## 5.1 Fitting models

Since the different approaches can identify complementary sets of regulatory 
relationships, we recommend fitting other models implemented in the `fitModel()` function
in addition to running TRIPOD.
We typically run model fitting for top 1000 highly variable genes.
In this vignette, we fit models for four examples genes, *Neurog2*,
*Eomes*, *Olig2*, and *Sox10* for demonstration purposes.

```{r, eval = TRUE}
genes <- c("Neurog2", "Eomes", "Olig2", "Sox10")
```

We set the size of the window around transcripion start sites (TSS) to 200 kb.
For a given gene, ATAC peaks within the windows are considered as possible 
*cis*-regulatory regions.

```{r, eval = TRUE}
ext.upstream <- ext.downstream <- 2e5
```

We first invoke the `getXYMatrices()` function.
For a given gene, the function returns a list containing a set of 
$Y_g$, $X_t$, and $Y_j$. Here, we obtain a list of lists for the four genes.

```{r, eval = TRUE}
xymats.list <- bplapply(
  genes,
  getXYMatrices,
  ext.upstream = ext.upstream,
  transcripts.gr = transcripts.gr,
  peaks.gr = peaks.gr,
  metacell.rna = metacell.rna,
  metacell.peak = metacell.peak,
  peakxmotif = peakxmotif,
  motifxTF = motifxTF,
  metacell.celltype = metacell.celltype,
  metacell.celltype.col = metacell.celltype.col
)
names(xymats.list) <- genes
```

We now use the `fitModel()` function to fit models. The table below
summarizes coefficients estimated in the models and input arguments
to the functions including `getTrios()`, which generates a data frame of screen hits.
See [our manuscript](https://www.biorxiv.org/content/10.1101/2021.09.22.461437v1) for details.

| Model name     | Formula      | Coefficient name | Input argument    |
|:---------------|:-------------|:-----------------|:------------------|
| marginal       | Yg ~ Xt      | alpha.m          | marginal.Yg.Xt    |
| marginal       | Yg ~ Yj      | beta.m           | marginal.Yg.Yj    |
| marginal       | Yg ~ Xt:Yj   | gamma.m          | NA                |
| marginal       | Yj ~ Xt      | eta.m            | marginal.Yj.Xt    |
| conditional    | Yg ~ Xt + Yj | alpha.c          | conditional.on.Yj |
| conditional    | Yg ~ Xt + Yj | beta.c           | conditional.on.Xt |
| interaction    | Yg ~ Xt*Yj   | alpha.i          | NA                |
| interaction    | Yg ~ Xt*Yj   | beta.i           | NA                |
| interaction    | Yg ~ Xt*Yj   | gamma.i          | interaction       |
| TRIPOD         |              | coef.level1      | TRIPOD            |
| TRIPOD         |              | coef.level2      | TRIPOD            |

The following code performs marginal testing.

```{r, eval = TRUE, message = FALSE}
xymats.m.list <- bplapply(
	xymats.list,
  fitModel,
	model.name = "marginal"
)
names(xymats.m.list) <- genes
```

The following code performs conditional testing.

```{r, eval = TRUE, message = FALSE}
xymats.c.list <- bplapply(
	xymats.list,
  fitModel,
	model.name = "conditional"
)
names(xymats.c.list) <- genes
```

The following code fits the linear interaction model. 

```{r, eval = TRUE, message = FALSE}
xymats.i.list <- bplapply(
	xymats.list,
  fitModel,
	model.name = "interaction"
)
names(xymats.i.list) <- genes
```

Finally, we run TRIPOD matching either $X_t$ or $Y_j$ using the following code.

```{r, eval = TRUE}
# run TRIPOD matching Xt
xymats.tripod.Xt.list <- bplapply(
	xymats.list,
	fitModel,
	model.name = "TRIPOD",
	match.by = "Xt"
)
names(xymats.tripod.Xt.list) <- genes
# run TRIPOD matching Yj
xymats.tripod.Yj.list <- bplapply(
	xymats.list,
	fitModel,
	model.name = "TRIPOD",
	match.by = "Yj"
)
names(xymats.tripod.Yj.list) <- genes
```

The resultant lists contain estimates of coefficients and associated $p$-values.
We show the results focusing on the marginal models and 
TRIPOD matching $X_t$ for the *Sox10* gene.

```{r, eval = TRUE}
xymats.tripod.Xt <- xymats.tripod.Xt.list$Sox10
names(xymats.tripod.Xt)
```

The `xymats.tripod.Xt$coef.level2` is a matrix containing estimates of 
coefficients from the level 2 test.
The $(t, j)$ element is the result with $t$-th ATAC peak and the $j$-th TF.
Note that the majority of the entries are `NA` due to the absence of
high-scoring TF motifs in ATAC peak regions or to inestimable coefficients. 

## 5.2 Getting lists of hits

We select hits by Benjamini-Hochberg (BH) method with a certain FDR threshold.
We typically use a threshold of 0.01 and focus on hits with significantly
positive coefficients. In the current version, the FDR method is applied for
a given target gene for the ease of computation, although it is more accurate
to apply the method across all analyzed target genes.

```{r, eval = TRUE}
# set FDR < 0.01
fdr.thresh <- 0.01

# focus on positive sign
sign <- "positive"
```

The `getTrios()` function returns a data frame with columns containing 
the gene name, peak index ($i$ for the $i$-th peak), 
TF index ($j$ for the $j$-th TF), peak name, TF name, 
estimates of the coefficient, nominal $p$-value, and BH-adjusted $p$-value. 
Here, we only show the code for TRIPOD level 1 and 2 matching $X_t$.

```{r, eval = TRUE}
# TRIPOD level 1 matchin Xt
xymats.tX1.pos.df <- getTrios(
	xymats.list = xymats.tripod.Xt.list,
	fdr.thresh = fdr.thresh,
	sign = sign,
  model.name = "TRIPOD",
	level = 1
)
head(xymats.tX1.pos.df[order(xymats.tX1.pos.df$adj), ])
```

```{r, eval = TRUE}
# TRIPOD level 2 matching Xt
xymats.tX2.pos.df <- getTrios(
	xymats.list = xymats.tripod.Xt.list,
	fdr.thresh = fdr.thresh,
	sign = sign,
  model.name = "TRIPOD",
	level = 2
)
head(xymats.tX2.pos.df[order(xymats.tX2.pos.df$adj), ])
```

The `getPeakGenePairs()` and `getTFGenePairs()` functions instead of 
`getTrios()` must be used for the `Yg ~ Xt` and `Yg ~ Yj` marginal models.

## 5.3 Visualizaing the results

We recommend visually inspecting the trio hits by generating scatter plots.
In this section, we demonstrate data visualization focusing on
the regulation of the *Sox10* gene by the Olig2 TF
via the peak chr15-79201373-79203205.

```{r, eval = TRUE}
gene.name <- "Sox10"
TF.name <- "Olig2"
peak.name <- "chr15-79201373-79203205"
```

First, we extract a `xymats` object for the target gene *Sox10*
from the `xymats.list` we have created.

```{r, eval = TRUE}
xymats <- xymats.list[[gene.name]]
```

We then invoke the `plotTFPeakGeneScatter()` function as follows.

```{r, eval = TRUE, fig.width = 6, fig.height = 9, fig.align = "center"}
par(pty = "s", mgp = c(2, 0.8, 0), mar = c(3, 3, 3, 1), mfrow = c(3, 2))
plotGenePeakTFScatter(
	xymats = xymats, peak.name = peak.name, TF.name = TF.name,
  to.plot = "marginal.Yg.Xt",
	cap.at.quantile = 0.02,
	metacell.celltype = metacell.celltype,
	metacell.celltype.col = metacell.celltype.col
)

plotGenePeakTFScatter(
	xymats = xymats, peak.name = peak.name, TF.name = TF.name,
  to.plot = "marginal.Yg.Yj",
	cap.at.quantile = 0.02,
	metacell.celltype = metacell.celltype,
	metacell.celltype.col = metacell.celltype.col
)

plotGenePeakTFScatter(
	xymats = xymats, peak.name = peak.name, TF.name = TF.name,
  to.plot = "marginal.Yj.Xt",
	cap.at.quantile = 0.02,
	metacell.celltype = metacell.celltype,
	metacell.celltype.col = metacell.celltype.col
)

plotGenePeakTFScatter(
	xymats = xymats, peak.name = peak.name, TF.name = TF.name,
  to.plot = "product",
	cap.at.quantile = 0.02,
	metacell.celltype = metacell.celltype,
	metacell.celltype.col = metacell.celltype.col
)

plotGenePeakTFScatter(
	xymats = xymats, peak.name = peak.name, TF.name = TF.name,
  to.plot = "conditional.on.Yj",
	cap.at.quantile = 0.02,
	metacell.celltype = metacell.celltype,
	metacell.celltype.col = metacell.celltype.col
)

plotGenePeakTFScatter(
	xymats = xymats, peak.name = peak.name, TF.name = TF.name,
  to.plot = "conditional.on.Xt",
	cap.at.quantile = 0.02,
	metacell.celltype = metacell.celltype,
	metacell.celltype.col = metacell.celltype.col
)
```

Additional visualization can provide
insights into the cell-type specificity of the regulation.
For this purpose, we first load the Seurat object, which can be generated
through a process described in [the preprocessing vignette](http://htmlpreview.github.io/?https://github.com/yharigaya/TRIPOD/blob/main/vignettes/preprocessing_e18.html).

```{r, include = FALSE, eval = TRUE, message = FALSE, warning = FALSE}
# for testing on a local machine only
file <- "e18.rds"
dir.data <- "~/Dropbox/TRIPOD/derived_data"
path <- file.path(dir.data, file)
e18 <- readRDS(path); rm(path)
```

```{r, eval = FALSE}
file <- "e18.rds"
path <- file.path("path/to/your/dir", file)
e18 <- readRDS(path); rm(path)
```

The following code visualizes RNA expression of the target gene,
chromatin accessibility of the ATAC peak region, and RNA expression of the gene
encoding the TF in UMAP, violin, and dot plots.

```{r, eval = TRUE, message = FALSE, warning = FALSE, fig.width = 9, fig.height = 3, fig.align = "center"}
plotGenePeakTFUMAP(object = e18, gene.name = gene.name,
	peak.name = peak.name, TF.name = TF.name,
	reduction = "umap.rna", order = TRUE,
	size.title = 12, size.axis = 12)
```

```{r, eval = TRUE, message = FALSE, warning = FALSE, fig.width = 9, fig.height = 4, fig.align = "center"}
plotGenePeakTFVln(object = e18, gene.name = gene.name,
	peak.name = peak.name, TF.name = TF.name,
	group.by = "celltype", size.title = 12, size.axis = 12)
```

```{r, eval = TRUE, message = FALSE, warning = FALSE, fig.width = 6, fig.height = 12, fig.align = "center"}
plotGenePeakTFDot(object = e18, gene.name = gene.name,
	peak.name = peak.name, TF.name = TF.name,
	group.by = "celltype", 
	size.title = 12, size.axis = 12, size.legend = 12, ncol = 1)
```

# 6 Predicting RNA expression using ATAC-seq data

The `performRNAPrediction()` function builds predictive models of gene expression
on chromatin accessibility of the vicinity region.
While this framework is not intended for genuinely prediction purposes since 
scRNA-seq data are required as predictors, it allows for examining 
whether TF RNA expression information improves 
the prediction of the target gene expression in a given dataset.

## 6.1 Fitting models

We demonstrate the predictive modeling focusing on the *Sox10* gene.
The following code fits a LASSO model with products of chromatin accessibility
and TF expression as predictors.
We utilize the `xymats` object we have created with a window size of 200 Kb 
(see Section 5).

```{r, eval = TRUE}
xymats.peak.tf.lasso <- performRNAPrediction(xymats = xymats, model.name = "peak.TF.LASSO",
	cap.at.quantile = 0.02, seed = NULL)
```

The model can be evaluated by Pearson correlation between the observed gene
expression and leave-one-out prediction.

```{r, eval = TRUE}
xymats.peak.tf.lasso$cor
```

For comparison, we build the gene activity model.

```{r, eval = TRUE}
xymats.gene.activity <- performRNAPrediction(xymats = xymats, model.name = "gene.activity",
	cap.at.quantile = 0.02, seed = NULL)
xymats.gene.activity$cor
```

We also preform LASSO on chromatin accessibility.

```{r, eval = TRUE}
xymats.peak.lasso <- performRNAPrediction(xymats = xymats, model.name = "peak.LASSO",
	cap.at.quantile = 0.02, seed = NULL)
xymats.peak.lasso$cor
```

## 6.2 Visualizaing the results

The models can be investigated by visualization as follows.

```{r, eval = TRUE, fig.width = 8, fig.height = 4, fig.align = "center"}
par(pty = "s", mgp = c(2, 0.8, 0), mar = c(3, 3, 3, 1), mfrow = c(1, 3))
plotRNAPrediction(
	xymats.gene.activity, metacell.celltype = metacell.celltype,
	metacell.celltype.col = metacell.celltype.col, cex.main = 0.9, cex.lab = 0.9)

plotRNAPrediction(
	xymats.peak.lasso, metacell.celltype = metacell.celltype,
	metacell.celltype.col = metacell.celltype.col, cex.main = 0.9, cex.lab = 0.9)

plotRNAPrediction(
	xymats.peak.tf.lasso, metacell.celltype = metacell.celltype,
	metacell.celltype.col = metacell.celltype.col, cex.main = 0.9, cex.lab = 0.9)
```

# 7 Identifying putative cell-type-specific trio regulatory relationships

In this section, we demonstrate the functionality of the TRIPOD package to infer 
cell types in which a trio regulatory relationship may be active 
by estimating the influence of a data point in models.
In the current version of the package, we only implement this framework with
the linear interaction model due to the large computational burdens of
performing the sampling procedure on the nonparametric models.

## 7.1 Preparation

We focus on the regulation of the *Neurog2*
gene by the Pax6 TF via the peak chr3-127639075-127643093, which gives
a significanly positive coefficient of the interaction term (see below).

```{r, eval = TRUE}
gene.name <- "Neurog2"
TF.name <- "Pax6"
peak.name <- "chr3-127639075-127643093"
```

We first obtain a `xymats` object.

```{r, eval = TRUE}
ext.upstream <- 2e5
xymats <- getXYMatrices(gene.name = gene.name, ext.upstream = ext.upstream,
  transcripts.gr = transcripts.gr, peaks.gr = peaks.gr,
  metacell.rna = metacell.rna, metacell.peak = metacell.peak,
  peakxmotif = peakxmotif, motifxTF = motifxTF,
  metacell.celltype = metacell.celltype, metacell.celltype.col = metacell.celltype.col
)
```

We then obtain indices for the TF and the peak.

```{r, eval = TRUE}
peak.num <- which(rownames(xymats$peakxmotif.g) == peak.name)
TF.num <- which(xymats$TF.g == TF.name)
```

Using the indices, we obtain the $Y_g$, $X_t$, and $Y_j$ vectors.

```{r, eval = TRUE}
Yg <- xymats$Yg
Xt <- xymats$Xt[, peak.num]
Yj <- xymats$Yj[, TF.num]
```

We then cap the values to eliminate outliers using `capValues()`.

```{r, eval = TRUE}
cap.at.quantile <- 0.02
Yg <- capValues(Yg, cap.at.quantile = cap.at.quantile)
Xt <- capValues(Xt, cap.at.quantile = cap.at.quantile)
Yj <- capValues(Yj, cap.at.quantile = cap.at.quantile)
```

We fit the linear interaction model to verify that the coefficient for the
interaction term (`gamma.i`) is significantly positive.

```{r, eval = TRUE, message = FALSE, warning = FALSE}
xymats.example <- fitModel(xymats = xymats, model.name = "interaction") 
c(xymats.example$gamma.i[peak.num, TF.num], xymats.example$pval.gamma.i[peak.num, TF.num])
```

## 7.2 Analyzing the influence of deleting one metacell at a time

To examine the influence of deleting one metacell, we compute Cook's D
using the `getCooksD()` function.
Please refer to the vignettes of the `oslrr` package [here](https://cran.r-project.org/web/packages/olsrr/vignettes/intro.html) and [here](https://cran.r-project.org/web/packages/olsrr/vignettes/influence_measures.html)
for details.

```{r, eval = TRUE}
d.list <- getCooksD(Yg, Xt, Yj, metacell.celltype)
p <- plotCooksD(d.list, metacell.celltype.col)
p
```

The following code computes DFFIT.

```{r, eval = TRUE}
d.list <- getDFFIT(Yg, Xt, Yj, metacell.celltype)
p <- plotDFFIT(d.list, metacell.celltype.col)
p
```

## 7.3 Analyzing the influence of deleting metacells of a given cell type at a time

To infer the cell type in which the trio regulatory relationship may be active,
we examine the influence of deleting metacells of each of the cell types
simultaneously using the `deleteCelltype()` function. The following code
obtain a matrix containing sampling $p$-values with the alternative hypothesis
that the deletion of metacells of a given cell type decreases the coefficient
in the regression model. That is, a small $p$-value indicates that the cell type
positively affects the coefficient.

```{r, eval = TRUE}
delta.coeff.pval <- deleteCellType(Yg = Yg, Xt = Xt, Yj = Yj,
	metacell.celltype = metacell.celltype, alternative = "greater")
```

The following code generates a plot of the $p$-values for the coefficient of
the interaction term.

```{r, eval = TRUE}
ordered.celltype <- c("Radial glia", "Neuroblast", "GABAergic",
	"Glutamatergic", "Glioblast", "Oligodendrocyte", "Cajal-Retzius")
p <- plotInfluenceForCellType(delta.coeff.pval = delta.coeff.pval, coefficient = "Xt:Yj",
	threshold = 0.01,
	metacell.celltype = metacell.celltype, metacell.celltype.col = metacell.celltype.col,
	ordered.celltype = ordered.celltype)
p
```

The above result implies that the regulation of the *Neurog2*
gene by the Pax6 TF via the peak chr3-127639075-127643093 may occur in
neuroblasts and glioblasts.

## 7.4 Analyzing the influence of deleting metacells of related cell types

The TRIPOD package also provides a functionality to analyze the influence of
deleting metacells of related cell types simultaneously. The first step
in this process is to obtain a tree structure of the cell types 
by hierarchical clustering. In this analysis, we use the RNA modality.

```{r, eval = TRUE}
dend <- getDendrogram(metacell.matrix = metacell.rna,
	metacell.celltype = metacell.celltype)
```

We then obtain sampling $p$-values for subtrees in the tree structure.

```{r, eval = TRUE}
delta.coeff.pval <- deleteBranch(Yg = Yg, Xt = Xt, Yj = Yj, dend = dend,
	alternative = "greater", metacell.celltype = metacell.celltype)
plotTree(dend = dend, delta.coeff.pval = delta.coeff.pval,
	metacell.celltype = metacell.celltype, coefficient = "Xt:Yj",
	metacell.celltype.col = metacell.celltype.col, threshold = 0.01)
```

The above result again implies that the regulation may occur in neuroblasts
and glioblasts. Although the significant results are restricted to leaf nodes
for this trio from the mouse embryonic btain data, we have observed examples 
where nodes in branches are consistently significant, particularly in the
PBMC dataset (see [our manuscript](https://www.biorxiv.org/content/10.1101/2021.09.22.461437v1)).

## 7.5 Visualizing the cell-type specificity of transcriptional regulation

UMAP plots of influential $p$-values can be used to visualize the cell-type specificity
of a given trio regulatory relationship.

To perform the visualization, it is necessary to obtain neighboring metacells for
each of the metacells first. The `getMetacellNeighbor()` function, which takes
a Seurat object and returns k nearest neighbors for each point, can be used for
this purpose. In the following code, we use the `e18` object, 
which has been read into the workspace in section 5.3, and set $k$ to 4.

```{r, eval = TRUE}
reduction <- "umap.rna"
metacell.neighbors <- getMetacellNeighbor(object = e18, reduction = reduction,
	cluster.name = "seurat_clusters", k = 4)
```

As before, we obtain sampling $p$-values for each group of metacells.
Note that the goal of this procedure is visualization and that the sampling 
$p$-values themselves may or may not be informative.

```{r, eval = TRUE}
delta.coeff.pval <- deleteNeighbor(Yg = Yg, Xt = Xt, Yj = Yj,
	metacell.celltype = metacell.celltype, metacell.neighbors = metacell.neighbors,
	alternative = "greater")
p <- plotInfluenceForNeighbor(delta.coeff.pval, coefficient = "Xt:Yj",
	metacell.celltype = metacell.celltype,
	metacell.celltype.col = metacell.celltype.col, threshold = 0.01)
p
```

The following code visualizes the sampling $p$-values for the coefficient of
the interaction term.

```{r, eval = TRUE, fig.width = 5, fig.height = 5, fig.align = "center"}
p <- plotFeatureForNeighbor(delta.coeff.pval = delta.coeff.pval, coefficient = "Xt:Yj",
	object = e18, reduction = reduction, label = FALSE)
p
```

# Session information

```{r}
sessionInfo()
```
